File: app/src/main/java/com/example/habitpalette/CreateHabitActivity.java
package com.example.habitpalette;

import android.app.DatePickerDialog;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.SeekBar;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CreateHabitActivity extends AppCompatActivity {
    private EditText mEditTextName;
    private EditText mEditTextStartDate;
    private RadioGroup mRadioGroupColor1;
    private RadioGroup mRadioGroupColor2;
    private CreateHabitPeriodSeekBar mSeekBarPeriod;
    private Button mButtonCreate;
    private String mHabitName;
    private Date mHabitStartDate;
    private int mHabitPeriod = 0;
    private int mHabitColor;
    private ArrayList<String> mArrayListPeriodType;
    private Date mCurrentDate;
    private CurrentHabitRecyclerItem mNewHabitItem;

    Calendar myCalendar = Calendar.getInstance();

    DatePickerDialog.OnDateSetListener myDatePicker = (view, year, month, dayOfMonth) -> {
        myCalendar.set(Calendar.YEAR, year);
        myCalendar.set(Calendar.MONTH, month);
        myCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
        try {
            updateLabel();
        } catch (ParseException e) {
            e.printStackTrace();
        }
    };

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_habit);

        mEditTextName = (EditText) findViewById(R.id.edit_text_habit_name);
        mEditTextStartDate = (EditText) findViewById(R.id.edit_text_habit_start_date);
        mSeekBarPeriod = (CreateHabitPeriodSeekBar) findViewById(R.id.seek_bar_habit_period);
        mRadioGroupColor1 = (RadioGroup) findViewById(R.id.radio_group_habit_color_row1);
        mRadioGroupColor2 = (RadioGroup) findViewById(R.id.radio_group_habit_color_row2);
        mButtonCreate = (Button) findViewById(R.id.button_habit_create);

        mEditTextName.addTextChangedListener(textWatcher);

        mEditTextStartDate.setOnClickListener(v -> new DatePickerDialog(CreateHabitActivity.this, myDatePicker, myCalendar.get(Calendar.YEAR), myCalendar.get(Calendar.MONTH), myCalendar.get(Calendar.DAY_OF_MONTH)).show());
        mEditTextStartDate.setInputType(InputType.TYPE_NULL);
        mEditTextStartDate.addTextChangedListener(textWatcher);

        mHabitColor = R.color.pink;
        mRadioGroupColor1.setOnCheckedChangeListener(colorSelected1);
        mRadioGroupColor2.clearCheck();
        mRadioGroupColor2.setOnCheckedChangeListener(colorSelected2);

        mSeekBarPeriod.setRangeCount(4);
        mArrayListPeriodType = new ArrayList<>();
        mArrayListPeriodType.add("10일");
        mArrayListPeriodType.add("30일");
        mArrayListPeriodType.add("50일");
        mArrayListPeriodType.add("100일");
        mSeekBarPeriod.setStrings(mArrayListPeriodType);

        mButtonCreate.setOnClickListener(createHabit);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if(mButtonCreate.isEnabled()){
            mButtonCreate.setTextColor(getResources().getColor(R.color.white));
            mButtonCreate.requestLayout();
        }else{
            mButtonCreate.setTextColor(getResources().getColor(R.color.black));
            mButtonCreate.requestLayout();
        }
    }

    private final TextWatcher textWatcher = new TextWatcher() {
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        public void onTextChanged(CharSequence s, int start, int before, int count) {
            String name = mEditTextName.getText().toString().trim();
            String startDate = mEditTextStartDate.getText().toString().trim();
            mButtonCreate.setEnabled(!(name.length()==0)&& !(startDate.length()==0));
        }
        public void afterTextChanged(Editable s) {
        }
    };

    private void updateLabel() throws ParseException {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy.MM.dd", Locale.KOREA);

        Date today = new Date();
        mCurrentDate = sdf.parse(sdf.format(today));
        mHabitStartDate = sdf.parse(sdf.format(myCalendar.getTime()));
        int compare = mCurrentDate.compareTo(mHabitStartDate);

        if (compare <= 0) {
            mEditTextStartDate.setText(sdf.format(myCalendar.getTime()));
        } else {
            Toast.makeText(CreateHabitActivity.this, "날짜 다시 선택", Toast.LENGTH_SHORT).show();
            mEditTextStartDate.setText("");
        }
    }

    View.OnClickListener createHabit = v -> {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy.MM.dd", Locale.KOREA);

        mHabitName=mEditTextName.getText().toString();
        try {
            mHabitStartDate=sdf.parse(sdf.format(mHabitStartDate.getTime()));
        } catch (ParseException e) {
            e.printStackTrace();
        }
        mHabitPeriod = mSeekBarPeriod.getProgress();
        Toast.makeText(CreateHabitActivity.this, mHabitName+mHabitStartDate+" "+Integer.toString(mHabitColor)+" "+Integer.toString(mHabitPeriod), Toast.LENGTH_SHORT).show();
    };

    RadioGroup.OnCheckedChangeListener colorSelected1 = new RadioGroup.OnCheckedChangeListener() {
        @Override
        public void onCheckedChanged(RadioGroup group, int checkedId) {
            if (checkedId != -1) {
                mRadioGroupColor2.setOnCheckedChangeListener(null);
                mRadioGroupColor2.clearCheck();
                mRadioGroupColor2.setOnCheckedChangeListener(colorSelected2);
            }
            if (checkedId == R.id.button_habit_color_pink) {
                mHabitColor = R.color.pink;
            } else if (checkedId == R.id.button_habit_color_red) {
                mHabitColor = R.color.red;
            } else if (checkedId == R.id.button_habit_color_orange) {
                mHabitColor = R.color.orange;
            } else if (checkedId == R.id.button_habit_color_yellow) {
                mHabitColor = R.color.yellow;
            } else if (checkedId == R.id.button_habit_color_green) {
                mHabitColor = R.color.green;
            }
        }
    };

    RadioGroup.OnCheckedChangeListener colorSelected2 = new RadioGroup.OnCheckedChangeListener() {
        @Override
        public void onCheckedChanged(RadioGroup group, int checkedId) {
            if (checkedId != -1) {
                mRadioGroupColor1.setOnCheckedChangeListener(null);
                mRadioGroupColor1.clearCheck();
                mRadioGroupColor1.setOnCheckedChangeListener(colorSelected1);
            }
            if (checkedId == R.id.button_habit_color_light_green_blue) {
                mHabitColor = R.color.light_green_blue;
            } else if (checkedId == R.id.button_habit_color_blue) {
                mHabitColor = R.color.blue;
            } else if (checkedId == R.id.button_habit_color_light_blue_purple) {
                mHabitColor = R.color.light_blue_purple;
            } else if (checkedId == R.id.button_habit_color_purple) {
                mHabitColor = R.color.purple;
            } else if (checkedId == R.id.button_habit_color_gray) {
                mHabitColor = R.color.gray;
            }
        }
    };
}

File: app/src/main/java/com/example/habitpalette/ui/habit/CreateHabitActivity.kt
package com.example.habitpalette.ui.habit

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.view.View
import android.widget.Toast
import androidx.activity.viewModels
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.example.habitpalette.HabitPaletteApplication
import com.example.habitpalette.R
import com.example.habitpalette.data.model.Habit
import com.example.habitpalette.databinding.ActivityCreateHabitBinding
import com.example.habitpalette.ui.home.MainActivity
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.*


class CreateHabitActivity : AppCompatActivity() {

    // 1. View Model 설정
    private val viewModel: CreateViewModel by viewModels{
        CreateViewModelFactory((application as HabitPaletteApplication).repository)
    }
    private lateinit var binding: ActivityCreateHabitBinding

    @RequiresApi(Build.VERSION_CODES.O)
    var mStartDate: LocalDate = LocalDate.MIN

    @RequiresApi(Build.VERSION_CODES.O)
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // 2. View Binding 설정
        binding = ActivityCreateHabitBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // 3. btnBack intent 설정
        binding.btnBack.setOnClickListener { view: View? ->
            val intent = Intent(applicationContext, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
            startActivity(intent)
        }

        // 4. etName에 textChangedListener 추가
        binding.etName.addTextChangedListener(textWatcher)

        // 5. etStartDate에 DatePickerDialog 추가
        binding.etStartDate.setOnClickListener {
            val cal = Calendar.getInstance()

            val dateSetListener = OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
                mStartDate = LocalDate.parse("${year}-${if((monthOfYear+1).toString().length!=1) {(monthOfYear+1).toString()} else {"0"+((monthOfYear+1).toString())}}-${if(dayOfMonth.toString().length!=1) {dayOfMonth.toString()} else {
                    "0$dayOfMonth"
                }}", DateTimeFormatter.ISO_DATE)
                val nowDate = LocalDate.now()
                if(nowDate.isBefore(mStartDate) || nowDate.isEqual(mStartDate)){
                    binding.etStartDate.setText(mStartDate.format(DateTimeFormatter.ofPattern("yyyy.MM.dd")))
                }else{
                    Toast.makeText(this@CreateHabitActivity, "날짜 다시 선택", Toast.LENGTH_SHORT).show()
                    binding.etStartDate.setText("")
                }
            }
            DatePickerDialog(this, dateSetListener, cal.get(Calendar.YEAR),cal.get(Calendar.MONTH),cal.get(Calendar.DAY_OF_MONTH)).show()
        }
        binding.etStartDate.inputType = InputType.TYPE_CLASS_DATETIME
        binding.etStartDate.addTextChangedListener(MaskWatcher.buildCpf())

        // 6. sbPeriod에 날짜 list 추가
        binding.sbPeriod.setRangeCount(4)
        binding.sbPeriod.setStrings(arrayListOf("10일", "30일", "50일", "100일"))

        // 7. 색상 선택
    }

    @RequiresApi(Build.VERSION_CODES.O)
    override fun onStart() {
        super.onStart()

        // 1. btnCreate에 setOnClickListener 추가
        binding.btnCreate.setOnClickListener{
            val title = binding.etName.text.toString()
            val startDate = LocalDate.parse(binding.etStartDate.text.toString().replace(".", ""), DateTimeFormatter.BASIC_ISO_DATE)
            val duration = returnDuration(binding.sbPeriod.progress)
            val endDate = startDate.plusDays(duration.toLong())
            val isCompleted = false
            val color = ContextCompat.getColor(this, R.color.pink)
            val usersId = 0L
            val score = 0f
            val startDateText = binding.etStartDate.text.toString()
            val endDateText = endDate.format(DateTimeFormatter.ofPattern("yyyy.MM.dd"))

            val habit = Habit(null, title,startDateText, endDateText,duration,isCompleted,color, usersId,score)
            viewModel.createHabitItem(habit)

            val intent = Intent(applicationContext, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
            startActivity(intent)
        }
    }

    override fun onResume() {
        super.onResume()
        if (binding.btnCreate.isEnabled) {
            binding.btnCreate.setTextColor(ContextCompat.getColor(this@CreateHabitActivity, R.color.white))
            binding.btnCreate.requestLayout()
        } else {
            binding.btnCreate.setTextColor(ContextCompat.getColor(this@CreateHabitActivity, R.color.gray))
            binding.btnCreate.requestLayout()
        }
    }

    private fun returnDuration(position: Int): Int{
        val dayList = listOf(10,30,50,100)
        return dayList[position]
    }

    private val textWatcher: TextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            val name = binding.etName.text.toString().trim { it <= ' ' }
            binding.btnCreate.isEnabled = name.isNotEmpty()
        }

        override fun afterTextChanged(s: Editable) {}
    }
}

class MaskWatcher(private val mask: String) : TextWatcher {
    private var isRunning = false
    private var isDeleting = false

    override fun beforeTextChanged(charSequence: CharSequence, start: Int, count: Int, after: Int) {
        isDeleting = count > after
    }

    override fun onTextChanged(charSequence: CharSequence, start: Int, before: Int, count: Int) {}

    override fun afterTextChanged(editable: Editable) {
        if (isRunning || isDeleting) {
            return
        }
        isRunning = true
        val editableLength = editable.length
        if (editableLength < mask.length) {
            if (mask[editableLength] != '#') {
                editable.append(mask[editableLength])
            } else if (mask[editableLength - 1] != '#') {
                editable.insert(editableLength - 1, mask, editableLength - 1, editableLength)
            }
        }
        isRunning = false
    }

    companion object {
        fun buildCpf(): MaskWatcher {
            return MaskWatcher("####.##.##")
        }
    }
}